### PR TITLE
Remove  ngraph backends

### PR DESCRIPTION
**Pull request by ilyachur**
### Details:
 - Remove IE backend
 - Moved INTERPRETER backend to template plugin
 - OV core unit tests are using Template plugin
 - This change is needed to intoduce ov_runtime library

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: docs/template_plugin/backend/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+add_library(interpreter_backend STATIC EXCLUDE_FROM_ALL ${SRC})
+
+if(CMAKE_COMPILER_IS_GNUCXX)
+    ie_add_compiler_flags(-Wno-missing-declarations)
+    ie_add_compiler_flags(-Wno-sign-compare)
+endif()
+
+ie_faster_build(interpreter_backend UNITY)
+
+ie_add_api_validator_post_build_step(TARGET interpreter_backend)
```
Comment (ilya-lavrenov) [inline docs/template_plugin/backend/CMakeLists.txt:29]: not needed for static libraries

File: docs/template_plugin/backend/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+target_compile_definitions(interpreter_backend
+    PRIVATE
+        SHARED_LIB_PREFIX="${CMAKE_SHARED_LIBRARY_PREFIX}"
+        SHARED_LIB_SUFFIX="${IE_BUILD_POSTFIX}${CMAKE_SHARED_LIBRARY_SUFFIX}"
+)
+target_link_libraries(interpreter_backend PRIVATE ngraph::builder ngraph::reference
+                                     pugixml::static ov_shape_inference openvino::util openvino::itt)
```
Comment (ilya-lavrenov) [inline docs/template_plugin/backend/CMakeLists.txt:37]: why all these dependencies are needed for `interpreter_backend`?

File: docs/template_plugin/backend/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+target_include_directories(interpreter_backend PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>)
+
+file(GLOB_RECURSE all_backends_src "${CMAKE_CURRENT_SOURCE_DIR}/*.cpp" "${CMAKE_CURRENT_SOURCE_DIR}/*.hpp")
+add_clang_format_target(interpreter_backend_clang FOR_SOURCES ${all_backends_src})
+
+
+# developer package
+openvino_developer_export_targets(COMPONENT core TARGETS interpreter_backend)
```
Comment (ilya-lavrenov) [inline docs/template_plugin/backend/CMakeLists.txt:46]: if interpreter is not used in tests, this line can be removed

File: docs/template_plugin/backend/backend.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+#include "int_backend.hpp"
+#include "ngraph/file_util.hpp"
+#include "ngraph/util.hpp"
+#include "openvino/util/file_util.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+std::mutex runtime::Backend::m_mtx;
+std::string runtime::Backend::s_backend_shared_library_search_directory;
```
Comment (ilya-lavrenov) [inline docs/template_plugin/backend/backend.cpp:18]: is `s_backend_shared_library_search_directory` used?

File: docs/template_plugin/src/CMakeLists.txt [hunk 5]
```diff
@@ -27,18 +27,14 @@ target_include_directories(${TARGET_NAME} PRIVATE
 target_link_libraries(${TARGET_NAME} PRIVATE
     IE::inference_engine
     IE::inference_engine_transformations
+    interpreter_backend
+    ngraph::reference
```
Comment (ilya-lavrenov) [inline docs/template_plugin/src/CMakeLists.txt:31]: why `ngraph::reference` are needed?

File: inference-engine/tests/ngraph_helpers/ngraph_functions/CMakeLists.txt [hunk 6]
```diff
@@ -18,6 +18,7 @@ addIeTarget(
         LINK_LIBRARIES
             PUBLIC
                 ngraph
+                ngraph_reference
                 interpreter_backend
```
Comment (ilya-lavrenov) [inline inference-engine/tests/ngraph_helpers/ngraph_functions/CMakeLists.txt:22]: why `ngraph_reference` is here?

File: src/core/tests/CMakeLists.txt [hunk 7]
```diff
@@ -69,28 +67,29 @@ set(SRC
...
+    # op_eval/non_zero.cpp
+    # op_eval/roi_align.cpp
+    # op_eval/roi_pooling.cpp
+    # op_eval/round.cpp
+    # op_eval/softplus.cpp
+    # op_eval/split.cpp
+    # op_eval/swish.cpp
+    # op_eval/strided_slice.cpp
+    # op_eval/transpose.cpp
+    # op_eval/variadic_split.cpp
```
Comment (ilya-lavrenov) [inline src/core/tests/CMakeLists.txt:92]: do we need to migrate these tests to new infra?

File: src/core/tests/CMakeLists.txt [hunk 8]
```diff
@@ -567,10 +519,10 @@ endif()
 target_link_libraries(ov_core_unit_tests PRIVATE ngraph_test_util
                                         engines_test_util
                                         ngraph::builder
+                                        ngraph_reference
```
Comment (ilya-lavrenov) [inline src/core/tests/CMakeLists.txt:522]: looks like it's a dependency of `interpreter_backend`? But why is it not a `PRIVATE`?

File: src/core/tests/engines_util/test_case.hpp [hunk 9]
```diff
@@ -12,18 +12,18 @@
...
 namespace test {
 std::shared_ptr<Function> function_from_ir(const std::string& xml_path, const std::string& bin_path = {});
 
-template <typename Engine, TestCaseType tct = TestCaseType::STATIC>
 class TestCase {
 public:
-    TestCase(const std::shared_ptr<Function>& function)
-        : m_engine{create_engine<Engine>(function, tct)},
-          m_function{function} {}
+    TestCase(const std::shared_ptr<Function>& function, const std::string& dev = "CPU") : m_function{function} {
```
Comment (ilya-lavrenov) [inline src/core/tests/engines_util/test_case.hpp:24]: why not `TEMPLATE`?

File: src/inference/CMakeLists.txt [hunk 10]
```diff
@@ -131,6 +131,11 @@ add_library(${TARGET_NAME}_obj OBJECT
             ${LIBRARY_HEADERS}
             ${PUBLIC_HEADERS})
 
+if(NOT BUILD_SHARED_LIBS)
+    target_compile_definitions(${TARGET_NAME}_obj PUBLIC OPENVINO_STATIC_LIBRARY)
+endif()
```
Comment (ilya-lavrenov) [inline src/inference/CMakeLists.txt:136]: is it needed in this PR?

File: src/core/tests/CMakeLists.txt (hunk 8 above)
Comment (ilyachur) [reply src/core/tests/CMakeLists.txt:522]: We use some headers from reference library

File: src/core/tests/CMakeLists.txt (hunk 7 above)
Comment (ilyachur) [reply src/core/tests/CMakeLists.txt:92]: Yes but at the current moment I enabled back all tests

File: docs/template_plugin/src/CMakeLists.txt (hunk 5 above)
Comment (ilyachur) [reply docs/template_plugin/src/CMakeLists.txt:31]: We use reference headers in the template plugin

File: docs/template_plugin/backend/backend.cpp (hunk 4 above)
Comment (ilyachur) [reply docs/template_plugin/backend/backend.cpp:18]: Removed

File: docs/template_plugin/backend/CMakeLists.txt (hunk 3 above)
Comment (ilyachur) [reply docs/template_plugin/backend/CMakeLists.txt:46]: Yes some private tests use Interpreter

File: docs/template_plugin/backend/CMakeLists.txt (hunk 2 above)
Comment (ilyachur) [reply docs/template_plugin/backend/CMakeLists.txt:37]: Fixed

File: docs/template_plugin/backend/CMakeLists.txt (hunk 1 above)
Comment (ilyachur) [reply docs/template_plugin/backend/CMakeLists.txt:29]: Fixed

File: src/core/tests/engines_util/test_case.hpp (hunk 9 above)
Comment (ilyachur) [reply src/core/tests/engines_util/test_case.hpp:24]: Fixed

File: src/inference/CMakeLists.txt (hunk 10 above)
Comment (ilyachur) [reply src/inference/CMakeLists.txt:136]: No we don't need it.

File: docs/template_plugin/backend/CMakeLists.txt [hunk 11]
```diff
@@ -5,51 +5,50 @@
...
         RUNTIME DESTINATION ${IE_CPACK_RUNTIME_PATH} COMPONENT tests OPTIONAL EXCLUDE_FROM_ALL
+        ARCHIVE DESTINATION ${IE_CPACK_RUNTIME_PATH} COMPONENT tests OPTIONAL EXCLUDE_FROM_ALL
         LIBRARY DESTINATION ${IE_CPACK_LIBRARY_PATH} COMPONENT tests OPTIONAL EXCLUDE_FROM_ALL)
 if(NOT BUILD_SHARED_LIBS)
-    install(TARGETS ngraph_backend
+    install(TARGETS interpreter_backend
             RUNTIME DESTINATION tests COMPONENT tests OPTIONAL EXCLUDE_FROM_ALL
+            ARCHIVE DESTINATION tests COMPONENT tests OPTIONAL EXCLUDE_FROM_ALL
             LIBRARY DESTINATION tests COMPONENT tests OPTIONAL EXCLUDE_FROM_ALL)
 endif()
```
Comment (ilya-lavrenov) [inline docs/template_plugin/backend/CMakeLists.txt:54]: looks like install rules can be removed (can be done separatelly)

File: src/core/tests/frontend/frontend_manager.cpp [hunk 12]
```diff
@@ -48,10 +83,9 @@ TEST(FrontEndManagerTest, testAvailableFrontEnds) {
     ASSERT_EQ(std::find(frontends.begin(), frontends.end(), "mock"), frontends.end());
 }
 
-TEST(FrontEndManagerTest, testMockPluginFrontEnd) {
+TEST(FrontEndManagerTest, DISABLED_testMockPluginFrontEnd) {
```
Comment (ilya-lavrenov) [inline src/core/tests/frontend/frontend_manager.cpp:86]: should be reverted I suppose

File: src/core/tests/frontend/shared/src/utils.cpp [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+#    include <windows.h>
+#    if defined(WINAPI_FAMILY) && !WINAPI_PARTITION_DESKTOP
+#        error "Only WINAPI_PARTITION_DESKTOP is supported, because of LoadLibrary[A|W]"
+#    endif
+#elif defined(__linux) || defined(__APPLE__)
+#    include <dlfcn.h>
+#endif
+
+std::string FrontEndTestUtils::find_ov_path() {
+#ifdef _WIN32
```
Comment (ilya-lavrenov) [inline src/core/tests/frontend/shared/src/utils.cpp:20]: can we use function from `ov::util`?

File: src/core/tests/frontend/shared/src/op_fuzzy.cpp [hunk 14]
```diff
@@ -67,7 +66,7 @@ void FrontEndFuzzyOpTest::runConvertedModel(const std::shared_ptr<ngraph::Functi
     auto modelFolder = getModelFolder(modelFile);
 
     // run test
-    auto testCase = test::TestCase<TestEngine>(function);
+    auto testCase = test::TestCase(function, "CPU");
```
Comment (ilya-lavrenov) [inline src/core/tests/frontend/shared/src/op_fuzzy.cpp:69]: why not template?

File: src/core/tests/eval.cpp [hunk 15]
```diff
@@ -166,32 +166,32 @@ TEST(eval, evaluate_dynamic_range_sum) {
...
-    auto result = backend->create_tensor();
-    auto cfun = backend->compile(fun);
-    cfun->call({result}, {p_start_val, p_stop_val, p_step_val, p1_val});
-    EXPECT_EQ(result->get_element_type(), element::f32);
-    EXPECT_EQ(result->get_partial_shape(), (PartialShape{3}));
-    auto result_val = read_vector<float>(result);
-    vector<float> seq{8.0f, 11.0f, 14.0f};
-    ASSERT_EQ(result_val, seq);
-}
+// TEST(eval, interpret_dynamic_range_sum) {
```
Comment (ilya-lavrenov) [inline src/core/tests/eval.cpp:169]: why?

File: src/core/tests/onnx/onnx_import_controlflow.in.cpp [hunk 16]
```diff
@@ -20,7 +20,7 @@ using namespace ngraph::test;
 
 static std::string s_manifest = "${MANIFEST}";
 
-using TestEngine = test::ENGINE_CLASS_NAME(${BACKEND_NAME});
+OPENVINO_SUPPRESS_DEPRECATED_START
```
Comment (ilya-lavrenov) [inline src/core/tests/onnx/onnx_import_controlflow.in.cpp:23]: what is suppressed in this file?

File: src/core/tests/type_prop/adaptive_avg_pool.cpp [hunk 17]
```diff
@@ -84,7 +84,7 @@ TEST(type_prop, adaptive_avg_pool_unsupported_input_shape) {
     EXPECT_THROW(make_shared<op::v8::AdaptiveAvgPool>(data, out_shape), NodeValidationFailure);
 }
 
-TEST(type_prop, adaptive_avg_pool_wrong_out_shape) {
+TEST(type_prop, DISABLED_adaptive_avg_pool_wrong_out_shape) {
```
Comment (ilya-lavrenov) [inline src/core/tests/type_prop/adaptive_avg_pool.cpp:87]: why?

**Review by ilya-lavrenov — APPROVED** [review]

File: CMakeLists.txt [hunk 18]
```diff
@@ -96,6 +96,11 @@ add_subdirectory(inference-engine/src)
 add_subdirectory(src)
 add_subdirectory(samples)
 add_subdirectory(inference-engine)
+
+# Enable interpreter backend
+if (ENABLE_TESTS OR ENABLE_TEMPLATE)
```
Comment (iefode) [inline CMakeLists.txt:101]: `AND` ?

Comment (ilyachur) [reply CMakeLists.txt:101]: OR

**Pull request merged**